funsor/terms.py

```diff
@@ -545,7 +550,6 @@ def __init__(self, data, dtype=None):
             data = type(dtype)(data)
         else:
             assert isinstance(dtype, str) and dtype == "real"
-            data = float(data)
```

Comment:
This took me quite a long while to track down. The issue here is that due to the behavior below, if we have a funsor expression like `u = 1 - v`, where v is a `Variable`, this would always give us an array of dtype `np.float64` instead of `np.int64` when we call `numpy.materialize`. The reason why this works for tensors is because PyTorch does not do dtype promotion. 

This change doesn't seem to break any existing tests, and seemed intuitive, but let me know if this is likely to cause other unintended consequences.

```python
>>> 1 - np.array([0, 1])
 array([1, 0])  # since 1 is int we get an long array

>>> 1. - np.array([0, 1])
 array([1., 0.])  # since 1. is float we get a double array 

>>> 1 - torch.tensor([0, 1])
 tensor([1, 0])  # long tensor

>>> 1. - torch.tensor([0, 1])
 tensor([1, 0])  # tensor type isn't promoted; still long tensor
```

Reply:
I'd rather keep these as Python floats here and do the downcasting to np.float32 in `binary_number_array`. The original intent was to avoid `int`s being stored as real `Number`s, because arithmetic would break in Python 2 user code. In funsor, there is no unbounded integer type, only bounded integers `bint(n)`. If you want to add a number to a long tensor, you should know the bound beforehand, like:
```py
1 + Tensor(torch.tensor(1, dtype=torch.long))  # should be a double tensor
Number(1) + Tensor(torch.tensor(1, dtype=torch.long))  # should be a real tensor
Number(1, 10) + Tensor(torch.tensor(1, dtype=torch.long), dtype=10)  # should be a long tensor dtype is bint(10)
```
Let's discuss subtleties though.

Reply:
That makes sense. I suppose one thing that we may need to be careful with then are methods that operate on the raw data type. e.g. in the first example, the underlying `.data` attribute will be holding a `torch.LongTensor`. e.g. I had to recast all the `.data` for indices to be `np.int64` in `Array.eager_subs` because the underlying numpy arrays were promoted to double even though the Array's dtype wasn't "real".